RTC beta (#1778)
tinymce/tinymce-docs: rtc/jwt-authentication.md

---
layout: default
title: RTC JWT Authentication
title_nav: JWT Authentication
description: Guide on how to setup JWT Authentication for RTC
keywords: jwt authentication
---

{% assign pluginname = "RTC" %}
{% assign plugincode = "rtc" %}
## Introduction

> RTC requires you to setup JSON Web Token (JWT) authentication. This is to ensure that only your authenticated users will be able to access and collaborate on documents.

Reviewer: ```suggestion
RTC requires you to setup JSON Web Token (JWT) authentication. This is to ensure that only authenticated users will be able to access and collaborate on documents.
```

> RTC requires you to setup JSON Web Token (JWT) authentication. This is to ensure that only your authenticated users will be able to access and collaborate on documents.

Reviewer 2: ```suggestion
RTC requires setting up JSON Web Token (JWT) authentication. This is to ensure that only authenticated users will be able to access and collaborate on documents.
```
Avoiding `you` and `your` again.

Reviewer 3: resolved in https://github.com/tinymce/tinymce-docs/pull/1778/commits/9cc200b9775691b5106c568717f3069b51dbb46e

> JWT is a standard authorization solution for web services and is documented in more detail at the [https://jwt.io/](https://jwt.io/) website. The guide aims to show how to setup JWT authentication for RTC.

Reviewer: ```suggestion
JWT is a standard authorization solution for web services and is documented in more detail at the [https://jwt.io/](https://jwt.io/) website. This guide aims to show how to setup JWT authentication for RTC.
```

{% include auth/private-public-key-pairs-for-tiny-cloud-services.md %}

{% include auth/jwt-provider-endpoint-url.md %}

## JWT requirements

{% include auth/jwt-supported-algorithms.md %}

### Claims

> These are like options/data you can send with the JWT token.

Reviewer: This info is now a table for RTC and a DT/DD for Drive - do we care that it's dissimilar?

Reviewer 3: table doesn't really work for Drive because of the URL, hence the definition list.

> These are like options/data you can send with the JWT token.

Reviewer 2: ```suggestion
Claims are additional data that can be sent as part of the JWT token. The RTC JWT token requires the following claims:
```

| Data | Optional or required | Description |
|---|:---:|---|
| `sub` | required | The unique user ID (i.e. if `sub` is the same for two clients, you should trust them as if they're the same user). |
| `exp` | required | The timestamp when the token expires. |

{% include auth/jwt-endpoint-setup-procedure.md %}

## Need help?

> We recommend reading up and trying to understand how JWT works; you need some necessary skills to implement RTC. This can be tricky, so if you need some help contact our support.

Reviewer: ```suggestion
Tiny recommends looking into how JWT works; some knowledge about JWT is necessary to implement RTC. This can be tricky, so if you need some help contact our support.
```
maybe?

## PHP token provider endpoint example

> This example uses the [Firebase JWT library](https://github.com/firebase/php-jwt) provided through the Composer dependency manager. The private key should be the private key that was generated through your {{site.accountpage}}.

Reviewer: ```suggestion
This example uses the [Firebase JWT library](https://github.com/firebase/php-jwt) provided through the Composer dependency manager. The private key should be the private key found at {{site.accountpage}}.
```
maybe?

### jwt.php

```php
<?php
require 'vendor/autoload.php';
use \Firebase\JWT\JWT;

header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept");

$privateKey = <<<EOD
-----BEGIN PRIVATE KEY-----
....
-----END PRIVATE KEY-----
EOD;

// NOTE: Before you proceed with the TOKEN, verify your users session or access.

$payload = array(
  "sub" => "123", // unique user ID string
  "exp" => time() + 60 * 10 // 10 minute expiration
);

try {
  $token = JWT::encode($payload, $privateKey, 'RS256');
  http_response_code(200);
  header('Content-Type: application/json');
  echo json_encode(array("token" => $token));
} catch (Exception $e) {
  http_response_code(500);
  header('Content-Type: application/json');
  echo $e->getMessage();
}
?>
```

### TinyMCE example with jwt.php Endpoint

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'rtc',
  rtc_token_provider: () => {
    return fetch('jwt.php', {
      method: 'POST'
    });
  }
});
```

## Node token provider endpoint example

> This example shows how to set up a Node.js express handler that produces the tokens. It requires you to install the Express web framework and the `jsonwebtoken` Node modules. For instructions on setting up a basic NodeJS Express server and adding {{site.productname}}, see: [Integrating TinyMCE into an Express JS App]({{site.baseurl}}/integrations/expressjs/).

Reviewer: ```suggestion
This example shows how to set up a Node.js express handler that produces the tokens. It requires you to install the Express web framework and the `jsonwebtoken` Node module. For instructions on setting up a basic NodeJS Express server and adding {{site.productname}}, see: [Integrating TinyMCE into an Express JS App]({{site.baseurl}}/integrations/expressjs/).
```

### /jwt

```js
const express = require('express');
const jwt = require('jsonwebtoken');
const cors = require('cors');

const app = express();
app.use(cors());

const privateKey = `
-----BEGIN PRIVATE KEY-----
....
-----END PRIVATE KEY-----
`;

app.post('/jwt', function (req, res) {
  // NOTE: Before you proceed with the TOKEN, verify your users' session or access.
  const payload = {
    sub: '123', // Unique user ID string
    exp: Math.floor(Date.now() / 1000) + (60 * 10) // 10 minutes expiration
  };

  try {
    const token = jwt.sign(payload, privateKey, { algorithm: 'RS256'});
    res.set('content-type', 'application/json');
    res.status(200);
    res.send(JSON.stringify({
      token: token
    }));
  } catch (e) {
    res.status(500);
    res.send(e.message);
  }
});

app.listen(3000);
```

### TinyMCE example with /jwt endpoint

```js
tinymce.init({
  selector: 'textarea',
  plugins: 'rtc',
  rtc_token_provider: () => {
    return fetch('/jwt', {
      method: 'POST'
    });
  }
});
```

### More configuration

> Once the JWT authentication has been set up, the RTC plugin can be configured further with the options shown on the [RTC configuration options page]({{site.baseurl}}/rtc/configuration/). Don't forget to change the JWT Claim's (user id, user name) to get those from your system.

Reviewer: ```suggestion
Once JWT authentication has been set up, the RTC plugin can be configured further using the options shown on the [RTC configuration options page]({{site.baseurl}}/rtc/configuration/). Don't forget to change the JWT Claim's (user id, user name) to get those from your system.
```
also, not sure what `to get those from your system.` means?

> If you need some help [submit a support request]({{site.supporturl}}).

Reviewer: ```suggestion
If you want help [submit a support request]({{site.supporturl}}).
```

---
layout: default
title: RTC JWT Authentication
title_nav: JWT Authentication
description: Guide on how to setup JWT Authentication for RTC
keywords: jwt authentication
---

## Introduction

> RTC requires you to setup JSON Web Token (JWT) authentication. This is to ensure that only your authenticated uses will be able to access and collaborate on documents.

Reviewer 3: Marking as resolved: Outdated before comment was added

> JWT is a standard authorization solution for web services and is documented in more detail at the [https://jwt.io/](https://jwt.io/) website. The guide aims to show how to setup JWT authentication for RTC.

Reviewer 3: Marking as resolved: Outdated before comment was added

## Private/public key pair

> {{site.cloudname}} services tokens use public/private RSA key pairs. The {{site.cloudname}} services only store the public key, allowing developers to have full control over the authentication.

Reviewer 3: Marking as resolved: Outdated before comment was added

> The private/public key pair can be created on your [{{site.companyname}} - {{site.accountpage}} page]({{site.accountpageurl}}), however we only store the public key on the {{site.accountpage}}. The private key should be downloaded and stored in your backend.

Reviewer 3: Marking as resolved: Outdated before comment was added

> **Important**: Keep the private key secure, do not commit files containing the key to public repositories or websites.

For information on generating an RSA key pair locally, see: [Creating a private/public key pair for RTC](#creatingaprivatepublickeypairforrtc).

## JWT provider URL

The easiest way to setup JWT authentication against {{site.cloudname}} services is to create a JWT provider endpoint. This endpoint takes a JSON HTTP POST request and produces a JSON result with the token that the service will then use for all the HTTP requests.

The following diagram explains the JWT call flow:

![JSON Web Token Call Flow]({{site.baseurl}}/images/jwt-call-flow.png "JSON Web Token Call Flow")

## JWT requirements

### Algorithm

> Our examples use, and we recommend RS256 algorithm. This is a list of supported ones: RS256, RS384, RS512, PS256, PS384, PS512

Reviewer 3: Marking as resolved: Outdated before comment was added

### Claims

> These are like options/data you can send with the JWT token.

Reviewer 3: Marking as resolved: Outdated before comment was added

* **sub** - _(required)_ The unique user ID (i.e. if `sub` is the same for two clients, you should trust them as if they're the same user)
* **exp** - _(required)_ The timestamp when the token expires

## JWT endpoint setup procedure

Follow these steps to set up your own JWT endpoint.

 1. Setup a JWT endpoint on your server, this could be a simple page using one of the examples below.
 2. Configure the `rtc_token_provider` to fetch the token from that endpoint.
 3. Make sure you copy the private key into the example code.
 4. You should be good to go now.

 > The JWT Endpoint should examine your systems sessions to verify your user has access to your system.

## Need help?

> We recommend reading up and trying to understand how JWT works; you need some necessary skills to implement RTC. This can be tricky if you need some help contact our support.

Reviewer 3: Marking as resolved: Outdated before comment was added
